src/net_processing.cpp

```diff
@@ -1264,6 +1264,9 @@ bool static ProcessHeadersMessage(CNode *pfrom, CConnman *connman, const std::ve
             if (nDoS > 0) {
                 LOCK(cs_main);
                 Misbehaving(pfrom->GetId(), nDoS);
+            } else if (!pfrom->fInbound && !pfrom->m_manual_connection && punish_invalid) {
+                // Don't keep outbound peers that are on an invalid chain
```

Comment:
This suggestion doesn't apply if you make the changes Cory and Andrew are asking, but otherwise it would be nice if this comment said something about why it avoids disconnecting when nDoS > 0.

Reply:
Reworked this and added an explanatory comment.